### PR TITLE
[5.2] Add updateOrInsert() method for query builder

### PR DESCRIPTION
**Pull request by themsaid**
Similar to `Eloquent/Builder::updateOrCreate($attributes, $values)`.

### CODE REVIEW

File: src/Illuminate/Database/Query/Builder.php [hunk 1]
```diff
@@ -1905,6 +1905,22 @@ public function update(array $values)
...
     /**
+     * Insert or update a record matching the attributes, and fill it with values.
+     *
+     * @param  array  $attributes
+     * @param  array  $values
+     * @return bool
+     */
+    public function updateOrInsert(array $attributes, array $values = [])
+    {
+        if (! $this->where($attributes)->first()) {
```
Comment (JosephSilber) [inline src/Illuminate/Database/Query/Builder.php]: Why not `exists()`?

File: tests/Database/DatabaseQueryBuilderTest.php [hunk 2]
```diff
@@ -1090,6 +1090,34 @@ public function testUpdateMethodRespectsRaw()
...
+        ]);
+
+        $builder->shouldReceive('where')->twice()->with(['email' => 'foo'])->andReturn(m::self());
+        $builder->shouldReceive('first')->once()->andReturn(m\Matcher\Any::class);
+        $builder->shouldReceive('take')->andReturnSelf();
+        $builder->shouldReceive('update')->once()->with(['name' => 'bar'])->andReturn(1);
+
+        $this->assertEquals(true, $builder->updateOrInsert(['email' => 'foo'], ['name' => 'bar']));
+    }
+
```
Comment (JosephSilber) [inline tests/Database/DatabaseQueryBuilderTest.php]: This whole test doesn't make much sense. This should be an integration test.

We have a lot of tests like these. Let's at least not add more of them.

File: src/Illuminate/Database/Query/Builder.php (hunk 1 above)
Comment (themsaid) [inline src/Illuminate/Database/Query/Builder.php]: good point, switched to `exists()`.

File: tests/Database/DatabaseQueryBuilderTest.php (hunk 2 above)
Comment (themsaid) [inline tests/Database/DatabaseQueryBuilderTest.php]: The test is concerned about checking if the update or insert methods are called based on the existence of a row, not sure if there's a better way to make such a test.

I'm still learning much looking at the framework tests, tried to find tests for `Eloquent/Builder::updateOrCreate()` but looks like this method is not covered and the closest alternative was `testFindOrFailMethodThrowsModelNotFoundException()` which mocks `first()` and `where()`.

If you can give me a hint about how to better test such method I'd be grateful.

**Pull request merged**